sharding/contracts/sharding_manager.sol

```diff
@@ -1,253 +1,260 @@
-    uint deposit;
-    // The collator's address
-    address addr;
+  event HeaderAdded(uint indexed shard_id, bytes32 parent_hash,
+                    bytes32 chunk_root, int128 period, int128 height,
+                    address proposer_address, uint proposer_bid,
+                    bytes proposer_signature);
+  event NotaryRegistered(address notary, uint pool_index);
+  event NotaryDeregistered(address notary, uint pool_index, uint deregistered_period);
+  event NotaryReleased(address notary, uint pool_index);
+  event ProposerRegistered(uint pool_index);
+  event ProposerDeregistered(uint index);
+  event ProposerReleased(uint index);
+
+  struct Notary {
+    uint deregisteredPeriod;
+    uint poolIndex;
+    bool deposited;
   }
-
-  struct CollationHeader {
-    bytes32 parentHash;
-    int score;
+  
+  struct Proposer {
+    uint deregisteredPeriod;
+    uint balances
   }
 
-  struct Receipt {
-    int shardId;
-    uint txStartgas;
-    uint txGasprice;
-    uint value;
-    bytes32 data;
-    address sender;
-    address to;
+  struct CollationHeader {
+    uint shardId;             // Number of the shard ID
+    bytes32 parentHash;       // Hash of the parent collation
+    bytes32 chunkRoot;        // Root hash of the collation body
+    uint period;              // Period which header should be included
+    uint height;              // Height of the collation
+    address proposerAddress;
+    uint proposerBid;     
+    bytes proposerSignature; 
   }
 
   // Packed variables to be used in addHeader
   struct HeaderVars {
     bytes32 entireHeaderHash;
     int score;
-    address collatorAddr;
+    address notaryAddr;
     bool isNewHead;
   }
 
-  // collatorId => Collators
-  mapping (int => Collator) public collators;
+  address[] public notaryPool;
+  Proposer[] public proposerPool;
+
+  // Notary registry (deregistered is 0 for not yet deregistered notaries)
+  mapping (address => Notary) public notaryRegistry;
+  mapping (address => Proposer) public proposerRegistry;
+  // shardId => (header_hash => tree root)
+  mapping (uint => mapping (bytes32 => bytes32)) public collationTrees;
   // shardId => (headerHash => CollationHeader)
   mapping (int => mapping (bytes32 => CollationHeader)) public collationHeaders;
-  // receiptId => Receipt
-  mapping (int => Receipt) public receipts;
   // shardId => headerHash
-  mapping (int => bytes32) shardHead;
-
-  // Number of collators
-  int public numCollators;
-  // Number of receipts
-  int numReceipts;
-  // Indexs of empty slots caused by the function `withdraw`
-  mapping (int => int) emptySlotsStack;
-  // The top index of the stack in empty_slots_stack
-  int emptySlotsStackTop;
-  // Has the collator deposited before?
-  mapping (address => bool) public isCollatorDeposited;
+  mapping (int => bytes32) shardead;
+
+  // Number of notaries
+  uint public notaryPoolLength;
+  // Indexes of empty slots caused by the function `withdraw`
```

Comment:
Do we still need this now that the withdraw function is gone?

Reply:
Still need it to track empty notary slot indices (in register/deregister function)
The comment is mislead, I will change it